packages/lit-html/src/lit-html.ts

```diff
@@ -319,15 +322,156 @@ const walker = d.createTreeWalker(
  * change in future pre-releases.
  */
 export abstract class Directive {
+  resolve(part: Part, props: Array<unknown>): unknown {
+    return this.update(part, props);
+  }
```

Comment:
In order to reuse the non-trivial `AttributePart` logic for running directives on all of the parts and concatenating everything together, I factored `_resolveValue` out of `_setValue`, so that we can call that during hydrate and on the server without committing the value to the DOM. However, as part of that we need to do the _server_ thing with directives, aka run its `render` not its `update`. In order to reuse the `AttributePart` logic, I made `Directive` have a patchable `resolve` method that defaults to running `update` but which can be patched by SSR to run `render` instead. This is maybe one of the more questionable changes. TBD

Reply:
Done (it's `@internal`, not private)